modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/index/IndexCreationElasticsearchConfigurationObserver.java

```diff
@@ -0,0 +1,144 @@
+ */
+
+package com.liferay.portal.search.elasticsearch7.internal.index;
+
+import com.liferay.osgi.service.tracker.collections.list.ServiceTrackerList;
+import com.liferay.osgi.service.tracker.collections.list.ServiceTrackerListFactory;
+import com.liferay.petra.io.Deserializer;
+import com.liferay.petra.io.Serializer;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.util.FileUtil;
+import com.liferay.portal.search.elasticsearch7.internal.configuration.ElasticsearchConfigurationObserver;
+import com.liferay.portal.search.elasticsearch7.internal.configuration.ElasticsearchConfigurationWrapper;
+import com.liferay.portal.search.elasticsearch7.internal.index.util.IndexFactoryCompanyIdRegistryUtil;
+import com.liferay.portal.search.spi.index.lifecycle.IndexLifecycleManager;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.OutputStream;
+
+import java.nio.ByteBuffer;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Bryan Engler
+ */
+@Component(service = {})
+public class IndexCreationElasticsearchConfigurationObserver
+	implements ElasticsearchConfigurationObserver {
+
+	@Override
+	public int compareTo(
+		ElasticsearchConfigurationObserver elasticsearchConfigurationObserver) {
+
+		return _elasticsearchConfigurationWrapper.compare(
+			this, elasticsearchConfigurationObserver);
+	}
+
+	@Override
+	public int getPriority() {
+		return 4;
+	}
+
+	@Override
+	public void onElasticsearchConfigurationUpdate() {
+		Bundle bundle = FrameworkUtil.getBundle(
+			IndexCreationElasticsearchConfigurationObserver.class);
+
+		_createIndexesOnElasticsearchConfigurationChange(
+			bundle.getBundleContext());
+	}
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_elasticsearchConfigurationWrapper.register(this);
+
+		_serviceTrackerList = ServiceTrackerListFactory.open(
+			bundleContext, IndexLifecycleManager.class);
+
+		_createIndexesOnElasticsearchConfigurationChange(bundleContext);
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		_elasticsearchConfigurationWrapper.unregister(this);
+	}
+
+	private void _createApplicationAndExternalIndexes() {
+		for (Long companyId :
+				IndexFactoryCompanyIdRegistryUtil.getCompanyIds()) {
+
+			for (IndexLifecycleManager indexLifecycleManager :
+					_serviceTrackerList) {
+
+				indexLifecycleManager.createIndex(companyId);
+			}
+		}
+	}
+
+	private void _createIndexesOnElasticsearchConfigurationChange(
+		BundleContext bundleContext) {
+
+		File dataFile = bundleContext.getDataFile(
+			"elasticsearch_configuration_state.data");
+
+		if (dataFile.exists()) {
```

Comment:
everything looks good to me, the only thing i want to confirm is if you researched adding these checks to this condition since ive seen it in other places?  
!StartupHelperUtil.isDBNew() && !StartupHelperUtil.isUpgrading()

Reply:
I didn't research much into it. Having pointed them out, !StartupHelperUtil.isDBNew() looks beneficial. Not sure about !StartupHelperUtil.isUpgrading() - why not have this run while upgrading?

I can look into it more.